src/main/java/bio/terra/service/kubernetes/KubePodListener.java

```diff
@@ -43,61 +52,93 @@ public void run() {
+
+                CoreV1Api api = new CoreV1Api();
+                try (Watch<V1Namespace> watch =
+                         Watch.createWatch(
+                             client,
+                             api.listNamespacedPodCall(namespace,
+                                 null,
+                                 null,
+                                 null,
+                                 null,
+                                 null,
+                                 5,
+                                 null,
+                                 null,
+                                 Boolean.TRUE,
+                                 null),
+                             new TypeToken<Watch.Response<V1Namespace>>() {
+                             }.getType())) {
+                    for (Watch.Response<V1Namespace> item : watch) {
+                        // If we are shutting down, we stop watching
+                        if (jobShutdownState.isShutdown()) {
+                            return;
+                        }
+                        // Reset retry if the watch worked
+                        consecutiveRetryCount = 0;
+                        retryWait = WATCH_INITIAL_WAIT;
+
+                        String operation = item.type;
+                        String podName = item.object.getMetadata().getName();
+                        logger.info(String.format("%s : %s", operation, podName));
 
-                    String operation = item.type;
-                    String podName = item.object.getMetadata().getName();
-                    logger.info(String.format("%s : %s%n", operation, podName));
-
-                    if (StringUtils.contains(podName, API_POD_FILTER)) {
-                        if (StringUtils.equals(operation, "ADDED")) {
-                            logger.info("Added api pod: " + podName);
-                            podMap.put(podName, true);
-                        } else if (StringUtils.equals(operation, "DELETED")) {
-                            // TODO: call Stairway to inform it of a deleted pod
-                            Boolean deletedPod = podMap.get(podName);
-                            if (deletedPod != null && !deletedPod) {
-                                logger.info("Deleted api pod: " + podName);
-                                podMap.put(podName, false);
+                        if (StringUtils.contains(podName, API_POD_FILTER)) {
+                            if (StringUtils.equals(operation, "ADDED")) {
+                                logger.info("Added api pod: " + podName);
+                                podMap.put(podName, true);
+                            } else if (StringUtils.equals(operation, "DELETED")) {
+                                try {
+                                    logger.info("Attempting clean up of deleted stairway instance: " + podName);
+                                    stairway.recoverStairway(podName);
```

Comment:
If a pod gets deleted, all running pods will try to recover it about the same time -- am I following this right? Is that a problem if they step on each other, or just contention on the db that could be handled with retries?

Reply:
It works because competing pods will arbitrate on the Stairway db. It has retries for serialization issues.

I do have an open issue with the current implementation on my list. (Not yet converted to tickets). If multiple pods recovery, they may add duplicates to the work queue. That is OK, but could get pretty wasteful as the pod count gets higher.